### PR TITLE
Add Vespa and rework Document Indices

### PR DESCRIPTION
**Pull request by yuhongsun96**

### CODE REVIEW

File: backend/alembic/versions/8aabb57f3b49_restructure_document_indices.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+# revision identifiers, used by Alembic.
+revision = "8aabb57f3b49"
+down_revision = "5e84129c8be3"
+branch_labels = None
+depends_on = None
+
+
+def upgrade() -> None:
+    op.drop_table("chunk")
+    op.drop_index(
```
Comment (Weves) [inline backend/alembic/versions/8aabb57f3b49_restructure_document_indices.py]: Hmm, why is this here?

File: backend/alembic/versions/8aabb57f3b49_restructure_document_indices.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        "ix_index_attempt_latest_for_connector_credential_pair",
+        table_name="index_attempt",
+    )
+    op.execute("DROP TYPE IF EXISTS documentstoretype")
+
+
+def downgrade() -> None:
+    op.create_index(
```
Comment (Weves) [inline backend/alembic/versions/8aabb57f3b49_restructure_document_indices.py]: Same as above

File: backend/danswer/configs/app_configs.py [hunk 3]
```diff
@@ -62,20 +64,28 @@
 #####
 # DB Configs
 #####
+DOCUMENT_INDEX_NAME = "danswer_index"  # Shared by vector/keyword indices
+# Vespa is now the default document index store for both keyword and vector
+DOCUMENT_INDEX_TYPE = os.environ.get(
+    "DOCUMENT_INDEX_TYPE", DocumentIndexType.SPLIT.value
```
Comment (Weves) [inline backend/danswer/configs/app_configs.py:70]: is this intended? Seems to contradict the comment above

File: backend/danswer/datastores/qdrant/store.py [hunk 4]
```diff
@@ -81,32 +84,96 @@ def _build_qdrant_filters(
...
+        )
+        for update_request in update_requests:
+            for doc_id_batch in batch_generator(
+                items=update_request.document_ids,
+                batch_size=_BATCH_SIZE,
+            ):
+                chunk_ids = _get_points_from_document_ids(
+                    doc_id_batch, self.collection, self.client
+                )
+                self.client.set_payload(
```
Comment (Weves) [inline backend/danswer/datastores/qdrant/store.py:151]: Note: we will be potentially updating quite a lot more than `_BATCH_SIZE` points here. Is that intended?

File: backend/danswer/datastores/qdrant/store.py [hunk 5]
```diff
@@ -81,32 +84,96 @@ def _build_qdrant_filters(
...
+                    points=chunk_ids,
+                )
+
+    def delete(self, doc_ids: list[str]) -> None:
+        logger.info(f"Deleting {len(doc_ids)} documents from Qdrant")
+        for doc_id_batch in batch_generator(items=doc_ids, batch_size=_BATCH_SIZE):
+            chunk_ids = _get_points_from_document_ids(
+                doc_id_batch, self.collection, self.client
+            )
+            self.client.delete(
```
Comment (Weves) [inline backend/danswer/datastores/qdrant/store.py:163]: same comment as above

File: backend/danswer/datastores/vespa/app_config/schemas/danswer_chunk.sd [hunk 6]
```diff
@@ -0,0 +1,89 @@
+schema danswer_chunk {
```
Comment (Weves) [inline backend/danswer/datastores/vespa/app_config/schemas/danswer_chunk.sd:1]: very cool!

File: backend/danswer/datastores/vespa/store.py [hunk 7]
```diff
@@ -0,0 +1,402 @@
...
+            f"Unexpected fetch document by ID value from Vespa "
+            f"with error {doc_fetch_response.status_code}"
+        )
+
+    doc_fields = doc_fetch_response.json()["fields"]
+    allowed_users = doc_fields.get(ALLOWED_USERS)
+    allowed_groups = doc_fields.get(ALLOWED_GROUPS)
+    # Add group permission later, empty list gets saved/loaded as a null
+    if allowed_users is None:
+        raise RuntimeError(
```
Comment (Weves) [inline backend/danswer/datastores/vespa/store.py:78]: might be better to have this match the logic of the other datastores

File: backend/danswer/datastores/vespa/store.py [hunk 8]
```diff
@@ -0,0 +1,402 @@
...
+    already_existing_documents: set[str] = set()
+    for chunk in chunks:
+        document = chunk.source_document
+        (
+            cross_connector_document_metadata_map,
+            should_delete_doc,
+        ) = update_cross_connector_document_metadata_map(
+            chunk=chunk,
+            cross_connector_document_metadata_map=cross_connector_document_metadata_map,
+            doc_store_cross_connector_document_metadata_fetch_fn=partial(
```
Comment (Weves) [inline backend/danswer/datastores/vespa/store.py:141]: why is this a partial?

File: backend/danswer/datastores/vespa/store.py [hunk 9]
```diff
@@ -0,0 +1,402 @@
...
+                _get_vespa_document_cross_connector_metadata,
+            ),
+            index_attempt_metadata=index_attempt_metadata,
+        )
+
+        if should_delete_doc:
+            # Processing the first chunk of the doc and the doc exists
+            deletion_success = _delete_vespa_doc_chunks(document.id)
+            if not deletion_success:
```
Comment (Weves) [inline backend/danswer/datastores/vespa/store.py:150]: nit: prefer to fail loudly in case that we fail to delete (plus add retries to delete)

File: backend/danswer/datastores/vespa/store.py [hunk 10]
```diff
@@ -0,0 +1,402 @@
...
+    def update(self, update_requests: list[UpdateRequest]) -> None:
+        logger.info(
+            f"Updating {len(update_requests)} documents' allowed_users in Vespa"
+        )
+
+        json_header = {"Content-Type": "application/json"}
+
+        for update_request in update_requests:
+            if update_request.boost is None and update_request.allowed_users is None:
+                logger.error("Update request received but nothing to update")
```
Comment (Weves) [inline backend/danswer/datastores/vespa/store.py:302]: should `continue` here?

File: backend/danswer/datastores/vespa/store.py [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+                update_dict["fields"][BOOST] = update_request.boost
+            if update_request.allowed_users:
+                update_dict["fields"][ALLOWED_USERS] = update_request.allowed_users
+
+            for document_id in update_request.document_ids:
+                for doc_chunk_id in _get_vespa_chunk_ids_by_document_id(document_id):
+                    url = f"{DOCUMENT_ID_ENDPOINT}/{doc_chunk_id}"
+                    res = requests.put(url, headers=json_header, json=update_dict)
+
+                    if res.status_code != 200:
```
Comment (Weves) [inline backend/danswer/datastores/vespa/store.py:315]: also fail loudly here if update fails to avoid marking the delete as successful without actually updating all permissions

File: backend/danswer/datastores/vespa/store.py [hunk 12]
```diff
@@ -0,0 +1,402 @@
...
+                        logger.error(f"Failed to update document: {document_id}")
+
+    def delete(self, doc_ids: list[str]) -> None:
+        logger.info(f"Deleting {len(doc_ids)} documents from Vespa")
+        for doc_id in doc_ids:
```
Comment (Weves) [inline backend/danswer/datastores/vespa/store.py:320]: this should probably also use `_delete_vespa_doc_chunks`?

File: backend/danswer/configs/app_configs.py (hunk 3 above)
Comment (yuhongsun96) [reply backend/danswer/configs/app_configs.py:70]: Originally this PR was suppose to swap to vespa but its separate now

File: backend/danswer/datastores/qdrant/store.py (hunk 4 above)
Comment (yuhongsun96) [reply backend/danswer/datastores/qdrant/store.py:151]: _BATCH_SIZE is quite conservative, and this is a tiny request in terms of packet size so I figure it shouldn't be an issue. Plus this won't really be used anymore anyway...

File: backend/danswer/datastores/vespa/app_config/schemas/danswer_chunk.sd (hunk 6 above)
Comment (yuhongsun96) [reply backend/danswer/datastores/vespa/app_config/schemas/danswer_chunk.sd:1]: lmao "very cool" haha
